### ISSUE TITLE
Improve streaming search speed

### ISSUE THREAD

**Issue by futurist** [zed-industries/zed#4560]
### Check for existing issues

- [X] Completed

### Describe the feature

VSCode uses [ripgrep](https://github.com/BurntSushi/ripgrep) for code search and its speed is amazing.

Ref: https://github.com/microsoft/vscode-ripgrep

In a repo with the same keyword search, Zed is 10x slower than VSCode.

**Comment by pconerly** [comment]
+1 on this, zed search is slow in larger repos. When coding I have both a Zed and a Vscode windows open and I feel goofy for it, but zed search is too slow!

Looking at vscode, they use ripgrep and parallelize it. Might be something to try! https://github.com/microsoft/vscode/blob/ab67de56b6646dac0113a498a57a4dec89cf101b/src/vs/workbench/services/search/node/ripgrepTextSearchEngine.ts

**Comment by dabljues** [comment]
Same issue here. Also comparing to VS Code, I find myself having to global search in VS Code, while zed either hangs up or even just crashes from the search.

**Comment by eirikhm** [comment]
Chiming in here after Zed came back on my radar. A global search in our codebase takes 30+sec, and the client is pretty much unresponsive while searching. I have a temp directory with some data files (JSON files at ~800MB). Moving those out resolved it.

**Comment by dabljues** [comment]
Yep. Thought I gotta say that even with many files, VS Code handles the search. It may take a sec longer (though it's streamed so it's not a problem), but it works. Zed can just hang or take forever. So while gitignoring/removing big files fixes the issue - it actually doesn't. It masks the problem that the search is just really slow.

**Comment by yajo** [comment]
I'm curious about the difference you find between zed and ripgrep. The difference I find is **abysmal**:

https://github.com/user-attachments/assets/72c669a0-9172-4a0e-80d7-dd04811c8666

Zed: about 1 minute.
Ripgrep: about 1 millisecond.

**Comment by RemcoSmitsDev** [comment]
Yeah zed with ripgrep is faster but to note opening buffers and adding them to the multibuffer is still the slowest code path. But by using ripgrep for search we now only open buffers that belong to a result instead of opening all buffers for a worktree untill we hit a limit of 5k so we can search through them.

**Comment by RemcoSmitsDev** [comment]
Yeah zed with ripgrep is faster but to note opening buffers and adding them to the multibuffer is still the slowest code path. But by using ripgrep for search we now only open buffers that belong to a result instead of opening all buffers for a worktree untill we hit a limit of 5k so we can search through them.

**Comment by m-ar-c** [comment]
I am used to projectile in Emacs who is also returning search queries instantly, no matter the size of the project.

Maybe a solution to very fast search is in indexing projects ?

**Comment by somtooo** [comment]
looks like a or has been opened to fix this

**Comment by yajo** [comment]
> opening buffers and adding them to the multibuffer is still the slowest code path

I'm not sure about that... For me it still seems like searching is the slowest part. For example, this search only gives 4 results in 1 file, but still takes more than 1 minute in Zed:

https://github.com/user-attachments/assets/6a1e37ef-2de1-44d6-b25f-f1f8f848bb7d

**Comment by albertosantini** [comment]
Just to know the behind of scenes, you may read https://zed.dev/blog/nerd-sniped-project-search

**Comment by hferreiro** [comment]
I can confirm that this is no longer an issue for me.
